py-polars/tests/unit/dataframe/test_from_dict.py

```diff
@@ -0,0 +1,192 @@
+            "value": {0: "x", 1: "y", 2: "z"}.values(),
+        },
+        schema={
+            "value": pl.Utf8,
+            "other": pl.Float32,
+            "misc": pl.Int32,
+            "key": pl.Int8,
+        },
+    )
+    assert df4.columns == ["value", "other", "misc", "key"]
+    assert df4.to_dict(False) == {
+        "value": ["x", "y", "z"],
+        "other": [7.0, 8.0, 9.0],
+        "misc": [4, 5, 6],
+        "key": [1, 2, 3],
+    }
+    assert df4.schema == {
+        "value": pl.Utf8,
+        "other": pl.Float32,
+        "misc": pl.Int32,
+        "key": pl.Int8,
+    }
+
+    # mixed with struct cols
+    for df5 in (
+        pl.from_dict(
+            {"x": {"b": [1, 3], "c": [2, 4]}, "y": [5, 6], "z": "x"},
+            schema_overrides={"y": pl.Int8},
+        ),
+        pl.from_dict(
+            {"x": {"b": [1, 3], "c": [2, 4]}, "y": [5, 6], "z": "x"},
+            schema=["x", ("y", pl.Int8), "z"],
+        ),
+    ):
+        assert df5.rows() == [({"b": 1, "c": 2}, 5, "x"), ({"b": 3, "c": 4}, 6, "x")]
+        assert df5.schema == {
+            "x": pl.Struct([pl.Field("b", pl.Int64), pl.Field("c", pl.Int64)]),
+            "y": pl.Int8,
+            "z": pl.Utf8,
+        }
+
+    # mixed with numpy cols...
+    df6 = pl.DataFrame(
+        {"x": np.ones(3), "y": np.zeros(3), "z": 1.0},
+    )
+    assert df6.rows() == [(1.0, 0.0, 1.0), (1.0, 0.0, 1.0), (1.0, 0.0, 1.0)]
+
+    # ...and trigger multithreaded load codepath
+    df7 = pl.DataFrame(
+        {
+            "w": np.zeros(1001, dtype=np.uint8),
+            "x": np.ones(1001, dtype=np.uint8),
+            "y": np.zeros(1001, dtype=np.uint8),
+            "z": 1,
+        },
+        schema_overrides={"z": pl.UInt8},
+    )
+    assert df7[999:].rows() == [(0, 1, 0, 1), (0, 1, 0, 1)]
+    assert df7.schema == {
+        "w": pl.UInt8,
+        "x": pl.UInt8,
+        "y": pl.UInt8,
+        "z": pl.UInt8,
+    }
+
+    # misc generators/iterables
+    df9 = pl.DataFrame(
+        {
+            "a": iter([0, 1, 2]),
+            "b": (2, 1, 0).__iter__(),
+            "c": (v for v in (0, 0, 0)),
+            "d": "x",
+        }
+    )
+    assert df9.rows() == [(0, 2, 0, "x"), (1, 1, 0, "x"), (2, 0, 0, "x")]
+
+
+@pytest.mark.slow()
+def test_from_dict_with_scalars_mixed() -> None:
+    # a bit of everything
+    mixed_dtype_data: dict[str, Any] = {
+        "a": 0,
+        "b": 8,
+        "c": 9.5,
+        "d": None,
+        "e": True,
+        "f": False,
+        "g": time(0, 1, 2),
+        "h": date(2023, 3, 14),
+        "i": timedelta(seconds=3601),
+        "j": datetime(2111, 11, 11, 11, 11, 11, 11),
+        "k": "「趣味でヒーローをやっている者だ」",
+    }
+    # note: deliberately set this value large; if all dtypes are
+    # on the fast-path it'll only take ~0.03secs. if it becomes
+    # even remotely noticeable that will indicate a regression.
+    # TODO: This is now slow (~0.15 seconds). Needs to be looked into.
```

Comment:
Added a TO DO here. Apparently there is a regression here?

Reply:
I'll check that - that timing may only relate to release builds, as well as being M1-specific (I'm compiling something up now to see if there's a real/meaningful drift there or not)

Reply:
@stinodego: Good news, can confirm no regression here; it's just quite sensitive to being release/debug or not. A fresh release build shows 0.0260 secs for mw, whereas debug was ~0.14 secs - so it's all good.

Reply:
All right, will remove the TO DO when I am refactoring some tests later! Thanks for looking into this 👍

Reply:
No problem; I recognised it as something I looked at before ;)